### PR TITLE
Adaptive components

### PR DESCRIPTION
**Pull request by JoseAlba**
Added Adaptive column and container

Adaptive Container does the same things as a regular container but also adds two functionalities:
- show up for specific windows types. 
- Tells the responsive screen how many columns this container should use. 

For example lets say you use the adaptive column widget than its children have to be a list of adaptive container. By default the adaptive container takes one column space and can be seen in all screens. 

As you can see in the example below: 

We have a few containers that only show up on specific types of screens and you can see that the containers is responsive depending on how big the screen is.

By default the adaptive container and column is dependent on the breakpoints package.

Also modified the adaptive breakpoints package. Removed adaptive container from package and modified variable names so it makes more sense. 

https://adaptive-components.web.app/#/

### CODE REVIEW

File: adaptive_breakpoints/example/lib/main.dart [hunk 1]
```diff
@@ -16,43 +16,48 @@ class AdaptiveBreakpointsExample extends StatelessWidget {
       home: Scaffold(
         body: Column(
           children: [
-            AdaptiveContainer(
-              adaptiveConstraints: AdaptiveConstraints(xs: true),
+            _AdaptiveContainer(
+              adaptiveConstraints:
+              _AdaptiveConstraints(xs: true),
```
Comment (rami-a) [inline adaptive_breakpoints/example/lib/main.dart]: The format on this is odd, it should go on the line above right? Same problem throughout this example

File: adaptive_breakpoints/example/lib/main.dart [hunk 2]
```diff
@@ -64,3 +69,190 @@ class AdaptiveBreakpointsExample extends StatelessWidget {
...
+///
+/// The AdaptiveContainer does everything a normal container does but with
+/// adaptive constraints. For more information go to one of the links below.
+///
+/// https://api.flutter.dev/flutter/widgets/Container-class.html
+/// https://material.io/design/layout/responsive-layout-grid.html#breakpoints
+///
+/// This class is useful whenever you want a container to only be active in
+/// certain [AdaptiveWindowType].
+class _AdaptiveContainer extends StatelessWidget {
```
Comment (rami-a) [inline adaptive_breakpoints/example/lib/main.dart]: So just so I understand this correctly.

The `AdaptiveContainer` was moved from the breakpoints package to another package, and so this breakpoints demo just makes its own private one? Instead could you add a dependency to the `AdaptiveContainer` to this demo?

Comment (rami-a) [reply adaptive_breakpoints/example/lib/main.dart]: If its not easy to add as a dependency, then you can probably remove most/all of the documentation on the private one and update the comment on it to say that its the same as the public one. If that makes sense?

File: adaptive_breakpoints/test/adaptive_breakpoints_test.dart [hunk 3]
```diff
@@ -285,3 +285,190 @@ void main() {
...
+///
+/// The AdaptiveContainer does everything a normal container does but with
+/// adaptive constraints. For more information go to one of the links below.
+///
+/// https://api.flutter.dev/flutter/widgets/Container-class.html
+/// https://material.io/design/layout/responsive-layout-grid.html#breakpoints
+///
+/// This class is useful whenever you want a container to only be active in
+/// certain [AdaptiveWindowType].
+class _AdaptiveContainer extends StatelessWidget {
```
Comment (rami-a) [inline adaptive_breakpoints/test/adaptive_breakpoints_test.dart]: Hmm duplicating this container a 3rd time seems bad. Maybe you could adjust the tests for breakpoints to not require this?

File: adaptive_breakpoints/test/adaptive_breakpoints_test.dart [hunk 4]
```diff
@@ -285,3 +285,190 @@ void main() {
...
+        }
+      },
+    );
+  }
+}
+
+/// Used to see if a range of [AdaptiveWindowType] should be shown in the window.
+/// If the user sets one of the variables below to true than that window type
+/// should be shown within the [AdaptiveContainer].
+class _AdaptiveConstraints {
```
Comment (rami-a) [inline adaptive_breakpoints/test/adaptive_breakpoints_test.dart]: This is also duplicated, if we could avoid that, that would be preferred.

File: adaptive_components/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# adaptive_components
```
Comment (rami-a) [inline adaptive_components/README.md:1]: Should you update this with actual content?

File: adaptive_components/example/.firebase/hosting.YnVpbGQvd2Vi.cache [hunk 6]
```diff
@@ -0,0 +1,12 @@
+favicon.png,1600088871060,fcc7c4545d5b62ad01682589e6fdc7ea03d0a3b42069963c815c344b632eb5cf
```
Comment (rami-a) [inline adaptive_components/example/.firebase/hosting.YnVpbGQvd2Vi.cache]: Hmm this seems like the kind of file that should be gitignored

File: adaptive_components/example/lib/adaptive_column_example.dart [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+class AdaptiveColumnsExample extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      home: Scaffold(
+        body: AdaptiveColumn(
+          children: [
+            AdaptiveContaine(
```
Comment (rami-a) [inline adaptive_components/example/lib/adaptive_column_example.dart]: Typo in the `AdaptiveContainer`

```suggestion
            AdaptiveContainer(
```

File: adaptive_components/lib/src/adaptive_column.dart [hunk 8]
```diff
@@ -0,0 +1,156 @@
+// Copyright 2020, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:adaptive_breakpoints/adaptive_breakpoints.dart';
+import 'package:flutter/material.dart';
+import 'adaptive_container.dart';
+
+class AdaptiveColumn extends StatelessWidget {
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart:24]: Since this is exported with the package, it needs documentation.

File: adaptive_components/lib/src/adaptive_column.dart [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    this.gutter,
+    this.margin,
+    this.columns,
+    @required this.children,
+  }) : assert(children != null);
+
+  final double margin;
+  final double gutter;
+  final AdaptiveConstraintsColumn columns;
+  final List<AdaptiveContaine> children;
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]: These all need documentation

File: adaptive_components/lib/src/adaptive_column.dart [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        );
+      },
+    );
+  }
+}
+
+/// Used to see if a range of [AdaptiveWindowType] should be shown in the window.
+/// If the user sets one of the variables below to true than that window type
+/// should be shown within the [AdaptiveContainer].
+class AdaptiveConstraintsColumn {
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]: I'm confused about what this class does. I think its name is a bit unclear since I would think this is a special kind of `Column` widget but it seems to instead just be a type of data structure. I suggest renaming it and updating its documentation

File: adaptive_components/lib/src/adaptive_column.dart [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+            }(),
+          ),
+        );
+      },
+    );
+  }
+}
+
+/// Used to see if a range of [AdaptiveWindowType] should be shown in the window.
+/// If the user sets one of the variables below to true than that window type
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]:
```suggestion
/// If the user sets one of the variables below to true then that window type
```

Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]: It says `true` but the variables are `int`s, so something doesn't line up

File: adaptive_components/lib/src/adaptive_column.dart [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        BreakpointSystemEntry _entry = getBreakpointEntry(context);
+        final double _margin = margin ?? _entry.margin;
+        final double _gutter = gutter ?? _entry.gutter;
+        final int _totalColumns =
+            columns?.getAdaptiveConstraintsColumn(context) == 0
+                ? columns.getAdaptiveConstraintsColumn(context)
+                : _entry.columns;
+
+        return Container(
+          color: Colors.grey,
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]: Why is this set to grey? It seems like it shouldn't have a color

File: adaptive_components/lib/src/adaptive_column.dart [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+                  if (currentColumns < _totalColumns) {
+                    totalGutters++;
+                  } else {
+                    int gutters = 0;
+                    if (currentColumns > _totalColumns) {
+                      totalGutters--;
+                    }
+
+                    for (AdaptiveContaine r in row) {
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]:
```suggestion
                    for (AdaptiveContaine rowItem in row) {
```

File: adaptive_components/lib/src/adaptive_column.dart [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+          constraints: BoxConstraints(
+            minWidth: _entry.adaptiveWindowType.widthRangeValues.start,
+            maxWidth: _entry.adaptiveWindowType.widthRangeValues.end,
+          ),
+          child: Wrap(
+            runSpacing: 8.0,
+            children: () {
+              int currentColumns = 0;
+              int totalGutters = 0;
+              List<Widget> childs = [];
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]: Rename to children, and then use `this.children` to access the class property.
```suggestion
              List<Widget> children = [];
```

File: adaptive_components/lib/src/adaptive_container.dart [hunk 15]
```diff
@@ -61,7 +63,19 @@ class AdaptiveContainer extends StatelessWidget {
   ///
   /// This is used by the builder to see what type of screen the user wants this
   /// [AdaptiveContainer] to fit within.
-  final AdaptiveConstraints adaptiveConstraints;
+  AdaptiveConstraints adaptiveConstraints;
+
+  /// AdaptiveColumn is used with [AdaptiveColumn] and [AdaptiveRow] to represent
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_container.dart]: I don't think there is a class called `AdaptiveRow`?

File: adaptive_components/lib/src/adaptive_container.dart [hunk 16]
```diff
@@ -61,7 +63,19 @@ class AdaptiveContainer extends StatelessWidget {
...
+  /// the amount of columns that this widget will fill up within a certain [Flex]
+  /// range.
+  ///
+  /// By default the columns will only represent one column space. If you want
+  /// this content of this widget to be shown increase it. There can be at most
+  /// 12 columns per flex range.
+  ///
+  /// Learn more by visiting the Material website:
+  /// https://material.io/design/layout/responsive-layout-grid.html#breakpoints
+  final int adaptiveColumn;
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_container.dart]: I suggest renaming as this is a bit confusing. Does it represent the number of columns?

File: adaptive_components/test/adaptive_components_test.dart [hunk 17]
```diff
@@ -0,0 +1,5 @@
+import 'package:flutter_test/flutter_test.dart';
```
Comment (rami-a) [inline adaptive_components/test/adaptive_components_test.dart:5]: Delete this file if there not going to be any tests, otherwise add some tests

File: adaptive_breakpoints/example/lib/main.dart (hunk 2 above)
Comment (rami-a) [reply adaptive_breakpoints/example/lib/main.dart]: Actually, what I recommend is leave the breakpoints package alone in this PR, and then once the `Adaptive components` package is published, update the breakpoints package to remove the adaptivecontainer and have the examples simplified perhaps?

File: adaptive_breakpoints/example/lib/main.dart (hunk 1 above)
Comment (JoseAlba) [reply adaptive_breakpoints/example/lib/main.dart]: Done

File: adaptive_breakpoints/example/lib/main.dart (hunk 2 above)
Comment (JoseAlba) [reply adaptive_breakpoints/example/lib/main.dart]: I thought of this but one of the problems with this is that the adaptive components package imports breakpoints which mean it imports the container within that package. This is the reason the adaptive component's container doesn't have the r at the end to differentiate between the two objects. 

To overcome this I need to remove the container from breakpoints. Release 0.0.3 and then make sure that components use that version in order to correct the name.

I don't want breakpoints to be dependent on components. It seems like an unnecessary import.

File: adaptive_components/example/.firebase/hosting.YnVpbGQvd2Vi.cache (hunk 6 above)
Comment (JoseAlba) [reply adaptive_components/example/.firebase/hosting.YnVpbGQvd2Vi.cache]: Added

File: adaptive_components/example/lib/adaptive_column_example.dart (hunk 7 above)
Comment (JoseAlba) [reply adaptive_components/example/lib/adaptive_column_example.dart]: I can't change the name of this file until I create a new version of the breakpoints package.

File: adaptive_components/lib/src/adaptive_column.dart (hunk 11 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 13 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 14 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Done

File: adaptive_components/lib/src/adaptive_container.dart (hunk 15 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_container.dart]: Done

File: adaptive_components/lib/src/adaptive_container.dart (hunk 16 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_container.dart]: Yeah it does. This woud represent the amount of column space it would take up within the range of the screen. If it is an xs screen and the adaptiveColumn is set to 4 then the container would take up the entire screen

File: adaptive_components/test/adaptive_components_test.dart (hunk 17 above)
Comment (JoseAlba) [reply adaptive_components/test/adaptive_components_test.dart:5]: DOne

File: adaptive_components/lib/src/adaptive_column.dart (hunk 12 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Removed, was for testing.

File: adaptive_breakpoints/example/lib/main.dart (hunk 2 above)
Comment (rami-a) [reply adaptive_breakpoints/example/lib/main.dart]: Gotcha,

Perhaps this should still be split up then:

1. Update breakpoints to no longer expose `AdaptiveContainer`
2. Create components package with `AdaptiveContainer`
3. (Optional) Have just the breakpoints example pubspec depend on components (would this break the circular dependency?)

File: adaptive_components/lib/src/adaptive_column.dart [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+
+class AdaptiveColumn extends StatelessWidget {
+  const AdaptiveColumn({
+    this.gutter,
+    this.margin,
+    this.columns,
+    @required this.children,
+  }) : assert(children != null);
+
+  final double margin;
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart:53]: Add documentation for each property

File: adaptive_components/lib/src/adaptive_column.dart [hunk 19]
```diff
@@ -0,0 +1,161 @@
+// Copyright 2020, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:adaptive_breakpoints/adaptive_breakpoints.dart';
+import 'package:flutter/material.dart';
+import 'adaptive_container.dart';
+
+class AdaptiveColumn extends StatelessWidget {
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart:24]: This is still missing documentation

File: adaptive_components/lib/src/adaptive_column.dart (hunk 10 above)
Comment (rami-a) [reply adaptive_components/lib/src/adaptive_column.dart]: Still a bit confused on what this class does.

File: adaptive_components/lib/src/adaptive_column.dart [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+            }(),
+          ),
+        );
+      },
+    );
+  }
+}
+
+/// Used to see if a range of [AdaptiveWindowType] should be shown in the window.
+/// If the user sets one of the variables below to true then that window type
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]: The variables are ints, so they can't be set to true.

File: adaptive_components/example/lib/adaptive_column_example.dart [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+              child: Text('This is an extra small window'),
+            ),
+            AdaptiveContainer(
+              adaptiveConstraints: AdaptiveConstraints.s(),
+              adaptiveColumn: 2,
+              color: Colors.orange,
+              child: Text('This is a small window'),
+            ),
+            AdaptiveContainer(
+              adaptiveConstraints: AdaptiveConstraints.m(),
```
Comment (rami-a) [inline adaptive_components/example/lib/adaptive_column_example.dart]: I think this won't actually compile since you renamed these:
```suggestion
              adaptiveConstraints: AdaptiveConstraints.medium(),
```

File: adaptive_navigation/test/adaptive_navigation_test.dart [hunk 22]
```diff
@@ -1,41 +0,0 @@
-import 'package:flutter/material.dart';
```
Comment (rami-a) [inline adaptive_navigation/test/adaptive_navigation_test.dart]: Why is this file being deleted? It seems unrelated

File: adaptive_components/example/lib/adaptive_column_example.dart (hunk 21 above)
Comment (JoseAlba) [reply adaptive_components/example/lib/adaptive_column_example.dart]: Done

File: adaptive_navigation/test/adaptive_navigation_test.dart (hunk 22 above)
Comment (JoseAlba) [reply adaptive_navigation/test/adaptive_navigation_test.dart]: Unchanged this

File: adaptive_components/lib/src/adaptive_column.dart (hunk 20 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Removed

File: adaptive_components/lib/src/adaptive_column.dart (hunk 18 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart:53]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 19 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart:24]: Done

File: adaptive_components/example/lib/adaptive_column_example.dart [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+        body: AdaptiveColumn(
+          children: [
+            AdaptiveContainer(
+              adaptiveConstraints: AdaptiveConstraints.xsmall(),
+              adaptiveColumn: 2,
+              color: Colors.red,
+              child: Text('This is an extra small window'),
+            ),
+            AdaptiveContainer(
+              adaptiveConstraints: AdaptiveConstraints.small(),
```
Comment (clocksmith) [inline adaptive_components/example/lib/adaptive_column_example.dart]: These fields should probably drop the `adaptive` prefix, since they exist within an AdaptiveContainer

File: adaptive_components/example/lib/main.dart [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+                Navigator.of(context).pushReplacement(
+                  MaterialPageRoute(
+                    builder: (context) {
+                      return AdaptiveColumnsExample();
+                    },
+                  ),
+                );
+              },
+            ),
+            SizedBox(height: 16),
```
Comment (clocksmith) [inline adaptive_components/example/lib/main.dart]: nit: const

File: adaptive_components/example/lib/main.dart [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+                Navigator.of(context).pushReplacement(
+                  MaterialPageRoute(
+                    builder: (context) {
+                      return AdaptiveContainerExample();
+                    },
+                  ),
+                );
+              },
+            ),
+            SizedBox(height: 16),
```
Comment (clocksmith) [inline adaptive_components/example/lib/main.dart:56]: nit: const

File: adaptive_components/lib/src/adaptive_column.dart [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+  Widget build(BuildContext context) {
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        BreakpointSystemEntry _entry = getBreakpointEntry(context);
+        final double _margin = margin ?? _entry.margin;
+        final double _gutter = gutter ?? _entry.gutter;
+        final int _totalColumns = _entry.columns;
+
+        return Container(
+          width: MediaQuery.of(context).size.width - (_margin * 2),
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart]: nit: parenthesis not needed

File: adaptive_components/lib/src/adaptive_column.dart [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+///
+/// Learn more about the breakpoint system:
+///
+/// https://material.io/design/layout/responsive-layout-grid.html#breakpoints
+class AdaptiveColumn extends StatelessWidget {
+  /// Creates a vertical array of children. Going from left to right and then
+  /// top to bottom.
+  ///
+  /// To see an example visit:
+  /// https://adaptive-components.web.app/#/
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart:38]: I don't think this link shows the column in action

File: adaptive_components/lib/src/adaptive_column.dart [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+///
+/// So if the user has 6 adaptive container and each container represents two columns
+/// then the 6 adaptive container would all fit within one Row on a extra large screen
+/// because extra large screens have 12 columns per row.
+///
+/// Learn more about the breakpoint system:
+///
+/// https://material.io/design/layout/responsive-layout-grid.html#breakpoints
+class AdaptiveColumn extends StatelessWidget {
+  /// Creates a vertical array of children. Going from left to right and then
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart:25]: I think this description needs more. Maybe describing how its different from a regular `Column` in Flutter ?

File: adaptive_components/lib/src/adaptive_column.dart [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+  /// Represents the space between children.
+  final double gutter;
+  final List<AdaptiveContainer> children;
+
+  @override
+  Widget build(BuildContext context) {
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        BreakpointSystemEntry _entry = getBreakpointEntry(context);
+        final double _margin = margin ?? _entry.margin;
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart:75]: If the constructor params override the values in `_entry`, the docs should mention that

File: adaptive_components/lib/src/adaptive_column.dart [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+        final double _gutter = gutter ?? _entry.gutter;
+        final int _totalColumns = _entry.columns;
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart]: `_totalColumns` not needed

File: adaptive_components/lib/src/adaptive_column.dart (hunk 26 above)
Comment (clocksmith) [reply adaptive_components/lib/src/adaptive_column.dart]: If the margin is being accounted for in the `margin` property, why would the width need to be further adjusted with the _margin value?

File: adaptive_components/lib/src/adaptive_column.dart [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+                              _gutter *
+                                  (gutters == totalGutters &&
+                                          currentColumns > _totalColumns
+                                      ? 0
+                                      : totalGutters)) /
+                          _totalColumns *
+                          rowItem.adaptiveColumn;
+                      children.add(
+                        ConstrainedBox(
+                          constraints: BoxConstraints(
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart:121]: `BoxConstraints.tightFor` may be used here

File: adaptive_components/lib/src/adaptive_column.dart [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+                // within the adaptive constraints.
+                if (child.adaptiveConstraints
+                    .withinAdaptiveConstraint(context)) {
+                  row.add(child);
+                  currentColumns += child.adaptiveColumn;
+
+                  if (currentColumns < _totalColumns) {
+                    totalGutters++;
+                  } else {
+                    int gutters = 0;
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart]: prefer to declare a local var as close as possible to its use

File: adaptive_components/lib/src/adaptive_column.dart [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+                      totalGutters--;
+                    }
+
+                    for (AdaptiveContainer rowItem in row) {
+                      // maxWidth equals column width without margin and gutters
+                      // divided by the total number of adaptive containers.
+                      double maxWidth = (MediaQuery.of(context).size.width -
+                              _margin * 2 -
+                              _gutter *
+                                  (gutters == totalGutters &&
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart]: will gutters always be 0 here?

File: adaptive_components/lib/src/adaptive_column.dart [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+                      if (gutters < totalGutters && 1 < row.length) {
+                        children.add(
+                          SizedBox(
+                            width: _gutter,
+                            child: Container(
+                              color: Colors.grey,
+                            ),
+                          ),
+                        );
+                        gutters++;
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart]: After this is incremented, where is it used?

File: adaptive_components/lib/src/adaptive_column.dart [hunk 35]
```diff
@@ -0,0 +1,131 @@
...
+                      }
+                    }
+                    totalGutters = 0;
+                    currentColumns = 0;
+                    row = [];
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_column.dart]: If `clear()` is used, row can probably be final

File: adaptive_components/lib/src/adaptive_container.dart [hunk 36]
```diff
@@ -0,0 +1,242 @@
...
+            decoration: decoration,
+            foregroundDecoration: foregroundDecoration,
+            transform: transform,
+            clipBehavior: clipBehavior,
+            height: height,
+            margin: EdgeInsets.symmetric(horizontal: _margin),
+            child: child,
+          );
+        } else {
+          return LimitedBox(
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_container.dart:157]: Is this a way to build nothing without making a new widget in the tree, or is that not possible?

File: adaptive_components/lib/src/adaptive_container.dart [hunk 37]
```diff
@@ -0,0 +1,242 @@
...
+  /// Use the [color] property to specify a simple solid color.
+  ///
+  /// The [child] is not clipped to the decoration. To clip a child to the shape
+  /// of a particular [ShapeDecoration], consider using a [ClipPath] widget.
+  final Decoration decoration;
+
+  /// The decoration to paint in front of the [child].
+  final Decoration foregroundDecoration;
+
+  /// Empty space to surround the [decoration] and [child].
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_container.dart:125]: Is this only horizontal margin?

File: adaptive_components/lib/src/adaptive_container.dart [hunk 38]
```diff
@@ -0,0 +1,242 @@
...
+  ///
+  /// Defaults to [Clip.none].
+  final Clip clipBehavior;
+
+  @override
+  Widget build(BuildContext context) {
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        if (adaptiveConstraints.withinAdaptiveConstraint(context)) {
+          final double _margin = margin ?? 0.0;
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_container.dart]: nit: local param not needed

File: adaptive_components/example/lib/main.dart (hunk 25 above)
Comment (JoseAlba) [reply adaptive_components/example/lib/main.dart:56]: done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 26 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Remove width

File: adaptive_components/example/lib/main.dart (hunk 24 above)
Comment (JoseAlba) [reply adaptive_components/example/lib/main.dart]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 28 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart:25]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 29 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart:75]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 30 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 31 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart:121]: tightfor doesn't allow for min and max width. The min and max width comes from the breakpoints package. It is the range value.

File: adaptive_components/lib/src/adaptive_column.dart (hunk 33 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: No gutter is just outside the foreach loop. 

At the end of the for each loop it will be changed if the if statement is set to true.

File: adaptive_components/lib/src/adaptive_column.dart (hunk 32 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: changed to rowGutters. This widget represents the number of gutters in the current row.

File: adaptive_components/lib/src/adaptive_column.dart (hunk 35 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 34 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: gutters now rowGutters represents the amount of gutters in the row currently being added. 

While totalGutters represents all the amount of gutters to expect to see in that row. Once rowGutters is equal to totalGutters than we don't need to add any more gutters

File: adaptive_components/lib/src/adaptive_container.dart (hunk 37 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_container.dart:125]: done

File: adaptive_components/lib/src/adaptive_container.dart (hunk 38 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_container.dart]: done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 31 above)
Comment (clocksmith) [reply adaptive_components/lib/src/adaptive_column.dart:121]: tightFor can take in a width, and then sets the minWidth and maxWidth to that value

File: adaptive_components/lib/src/adaptive_container.dart [hunk 39]
```diff
@@ -0,0 +1,241 @@
...
+    this.color,
+    this.decoration,
+    this.foregroundDecoration,
+    this.margin,
+    this.transform,
+    this.height,
+    this.child,
+    this.clipBehavior = Clip.none,
+    this.adaptiveConstraints,
+    this.adaptiveColumn = 1,
```
Comment (clocksmith) [inline adaptive_components/lib/src/adaptive_container.dart]: is this columnWidth?

Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_container.dart]: No, it represents the number of columns it should consume. So if is 4 and there are 3 adaptvie columns in a xlarge screen than it the adaptive columns would take up the row. But if it was a small screen that the adaptive columns would appear as a regular column

File: adaptive_components/lib/src/adaptive_container.dart (hunk 36 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_container.dart:157]: I don't think so. But I think this is an efficient widget

File: adaptive_components/lib/src/adaptive_column.dart (hunk 27 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart:38]: Updated

File: adaptive_components/lib/src/adaptive_container.dart (hunk 39 above)
Comment (clocksmith) [reply adaptive_components/lib/src/adaptive_container.dart]: Maybe a better name would be columns or adaptiveColumns (although the adaptive prefix is not really necessary since we are in context of an AdaptiveContainer)

File: adaptive_components/lib/src/adaptive_column.dart (hunk 32 above)
Comment (clocksmith) [reply adaptive_components/lib/src/adaptive_column.dart]: If the only code consuming this var is the loop over the row below, then it should be declared there for better readability

File: adaptive_components/lib/src/adaptive_container.dart (hunk 36 above)
Comment (clocksmith) [reply adaptive_components/lib/src/adaptive_container.dart:157]: Ok, would using an empty `Container()` provide the same efficiency?

File: adaptive_components/example/lib/adaptive_column_example.dart (hunk 23 above)
Comment (JoseAlba) [reply adaptive_components/example/lib/adaptive_column_example.dart]: Should I change the name of the class as well. I thought having adaptive as a prefix to this package would help users in terms of autocomplete.

Comment (rami-a) [reply adaptive_components/example/lib/adaptive_column_example.dart]: I think this is what was meant. The class should keep the Adaptive prefix. Right @clocksmith ?
```suggestion
              constraints: AdaptiveConstraints.small(),
```

**Review by clocksmith — APPROVED** [review]
LGTM once API names are agreed on and changed if needed

File: adaptive_components/example/lib/main.dart [hunk 40]
```diff
@@ -0,0 +1,63 @@
...
```
Comment (rami-a) [inline adaptive_components/example/lib/main.dart:56]: Nit: Remove this since it will make the buttons off center

File: adaptive_components/lib/src/adaptive_column.dart [hunk 41]
```diff
@@ -0,0 +1,152 @@
...
+  /// double value will be dependent of the breakpoint entry of the current screen.
+  ///
+  /// Learn more at https://material.io/design/layout/responsive-layout-grid.html#breakpoints
+  final double gutter;
+
+  /// The List of [AdaptiveContainer]. Adaptive container neeeds to be used
+  /// because the widget has a column parameter. This parameter represents the
+  /// amount of columns this widget should incompass.
+  ///
+  /// By default it is set to 1.
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_column.dart]: Looks like the column parameter no longer exists. This comment needs to be updated. The default doesn't match the parameter type either

File: adaptive_components/lib/src/adaptive_container.dart [hunk 42]
```diff
@@ -0,0 +1,241 @@
...
+            decoration: decoration,
+            foregroundDecoration: foregroundDecoration,
+            transform: transform,
+            clipBehavior: clipBehavior,
+            height: height,
+            margin: margin,
+            child: child,
+          );
+        } else {
+          return LimitedBox(
```
Comment (rami-a) [inline adaptive_components/lib/src/adaptive_container.dart:157]: Hmm would just an empty `Container()` be cleaner than this? Also add a small comment explaining why this is needed.

**Review by rami-a — APPROVED** [review]
LGTM

File: adaptive_components/example/lib/main.dart (hunk 40 above)
Comment (JoseAlba) [reply adaptive_components/example/lib/main.dart:56]: Done

File: adaptive_components/lib/src/adaptive_column.dart (hunk 41 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_column.dart]: Done

File: adaptive_components/lib/src/adaptive_container.dart (hunk 42 above)
Comment (JoseAlba) [reply adaptive_components/lib/src/adaptive_container.dart:157]: Done. I am using the limitedBox since it is a more efficient widget to use.
